generate_uninstall/CMakeLists.txt

```diff
@@ -50,7 +53,10 @@ else ()
    ")
    install (FILES "${CMAKE_CURRENT_BINARY_DIR}/${UNINSTALL_PROG}" DESTINATION "${UNINSTALL_PROG_DIR}" PERMISSIONS ${UNINSTALL_PERMISSIONS})
 
-   add_custom_target (uninstall
-      COMMAND sh "${CMAKE_CURRENT_BINARY_DIR}/${UNINSTALL_PROG}"
-   )
```

Comment:
Suggest a different design:

```cmake 
if(NOT TARGET uninstall)
  add_custom_target(uninstall)
endif()
add_custom_target(mongo-c-driver-uninstall COMMAND [do the uninstall])
add_dependencies(uninstall mongo-c-driver-uninstall)
```

This way, we "add" to an existing `uninstall` target if it already defined (e.g. by a parent project).

Reply:
Wouldn't it be preferable to defer the responsibility of including C Driver files in the uninstall command to the CXX Driver (top-level project) rather than unconditionally hooking into it from the C Driver side? The CXX Driver may want to uninstall the C Driver as part of its uninstall command, but another top-level project (i.e. 3rd party that `add_subdirectory()`'s the C Driver) might not. As I understand, this would either unconditionally attach the C Driver to any existing `uninstall` command regardless if the parent project actually wants to include it or not, or force the parent project to hook into the C Driver's `uninstall` target instead.

Reply:
That's tricky, since the C driver unconditionally(?) adds its install rules, which the parent may not want, which would imply that it should also unconditionally(?) add its uninstall rules as well. It is possible to `EXCLUDE_FROM_ALL` on the `add_subdirectory()`, but I'm not sure if that affects the `install()` commands (the documentation doesn't mention it, but one would intuit that it *should*).

Reply:
A quick local test suggests that `EXCLUDE_FROM_ALL` _does_ exclude the subproject's install rules when running the `install` target in the parent project. This is consistent with the [GNU Make documentation](https://www.gnu.org/software/make/manual/make.html) for the `install` target:

> If possible, write the `install` target rule so that it does not modify anything in the directory where the program was built, provided ‘`make all`’ has just been done. This is convenient for building the program under one user name and installing it under another.

combined with [CMake documentation](https://cmake.org/cmake/help/latest/command/add_subdirectory.html) for `EXCLUDE_FROM_ALL`:

> If the `EXCLUDE_FROM_ALL` argument is provided then targets in the subdirectory will not be included in the ALL target of the parent directory by default, and will be excluded from IDE project files. Users must explicitly build targets in the subdirectory.

This is also consistent with [this comment](https://discourse.cmake.org/t/is-install-of-a-target-with-exclude-from-all-true-really-undefined/5666/13) by a CMake maintainer, although their remarks are not as confident as I would have hoped:

> Regarding the overarching theme of excluding things from the `ALL` target and then installing them, the thing to remember is that if a target is excluded from `ALL`, then you the user are responsible for ensuring it is built if you then do something that wants to install that target. The documentation for the `install()` command calls this undefined behavior, but that’s perhaps a bit strong. In reality, my preceding sentence is the real constraint, as far as I’m aware (but I haven’t checked how export sets affect the picture).

The documentation stating "undefined behavior" seems to have been changed from [CMake 3.17](https://cmake.org/cmake/help/v3.17/prop_tgt/EXCLUDE_FROM_ALL.html) to [CMake 3.18](https://cmake.org/cmake/help/v3.18/prop_tgt/EXCLUDE_FROM_ALL.html) (current wording):

> If a target has `EXCLUDE_FROM_ALL` set to true, it may still be listed in an `install(TARGETS)` command, but the user is responsible for ensuring that the target’s build artifacts are not missing or outdated when an install is performed.

Summary: there seems to be a lot more "should" qualifiers to this than one would expect or hope, but it _should_ behave as expected (`EXCLUDE_FROM_ALL` -> exclude from `all` target -> also exclude from `install` target).